File: leetcode/Leetcode50.py
class Solution:
    def myPow(self, x: float, n: int) -> float:
        def helper(n):
            if n == 0:
                return 1
            if n == 1:
                return x
            y = helper(n // 2)
            return y ** 2 if n & 1 == 0 else x * y ** 2

        return helper(n) if n >= 0 else 1.0/helper(-n)


if __name__ == '__main__':
    print(Solution().myPow(0.00001, 2147483647))

File: java/Leetcode50.java
public class Leetcode50 {
    public double myPow(double x, int n) {
        if(n >= 0){
            return helper(x, n);
        }
        else return 1/helper(x, -n);
    }

    public double helper(double x, int n){
        if(n == 0){
            return 1;
        }
        if(n == 1){
            return x;
        }
        if( (n & 1) == 0){
            return helper(x, n >>> 1) * helper(x, n >>> 1);
        }else{
            return x * helper(x, (n-1)>>> 1) * helper(x, (n-1)>>>1);
        }
    }

    public static void main(String[] args) {
        System.out.println(
                new Leetcode50().myPow(2.00000,
                        -2147483648)
        );
    }
}
